src/QsCompiler/Transformations/CallGraph/ConcreteCallGraphWalker.cs

```diff
@@ -0,0 +1,293 @@
+
+                    var referenceRange = Range.Zero;
+                    if (this.CurrentStatementOffset.IsValue
+                        && this.CurrentExpressionRange.IsValue)
+                    {
+                        referenceRange = this.CurrentStatementOffset.Item + this.CurrentExpressionRange.Item;
+                    }
+
+                    void AddEdge(QsSpecializationKind kind)
+                    {
+                        var called = new ConcreteCallGraphNode(identifier, kind, typeRes);
+                        var edge = new ConcreteCallGraphEdge(referenceRange);
+                        this.Graph.AddDependency(this.CurrentNode, called, edge);
+                        if (!this.RequestStack.Contains(called) && !this.ResolvedNodeSet.Contains(called))
+                        {
+                            this.RequestStack.Push(called);
+                        }
+                    }
+
+                    if (this.IsInCall)
+                    {
+                        if (this.HasAdjointDependency && this.HasControlledDependency)
+                        {
+                            AddEdge(QsSpecializationKind.QsControlledAdjoint);
+                        }
+                        else if (this.HasAdjointDependency)
+                        {
+                            AddEdge(QsSpecializationKind.QsAdjoint);
+                        }
+                        else if (this.HasControlledDependency)
+                        {
+                            AddEdge(QsSpecializationKind.QsControlled);
+                        }
+                        else
+                        {
+                            AddEdge(QsSpecializationKind.QsBody);
+                        }
+                    }
+                    else
+                    {
+                        // The callable is being used in a non-call context, such as being
+                        // assigned to a variable or passed as an argument to another callable,
+                        // which means it could get a functor applied at some later time.
+                        // We're conservative and add all possible kinds defined for the callable.
+                        foreach (var kind in this.GetSpecializationKinds(identifier))
+                        {
+                            AddEdge(kind);
+                        }
+                    }
+                }
+            }
+
+            private class NamespaceWalker : NamespaceTransformation<TransformationState>
+            {
+                public NamespaceWalker(SyntaxTreeTransformation<TransformationState> parent) : base(parent, TransformationOptions.NoRebuild)
+                {
+                }
+
+                public override QsGeneratorDirective OnGeneratedImplementation(QsGeneratorDirective directive)
+                {
+                    throw new ArgumentException("Encountered unresolved generated specialization while constructing concrete call graph.");
```

Comment:
Is this documented in the corresponding public method/class?

Reply:
The documentation in the relevant places has been updated to make it clear that the concrete call graph constructor expects there to be no Generated Implementation directives, and will error if one is encountered.